src/rpc/mining.cpp

```diff
@@ -942,17 +942,17 @@ static UniValue submitblock(const JSONRPCRequest& request)
     }
 
     bool new_block;
-    submitblock_StateCatcher sc(block.GetHash());
-    RegisterValidationInterface(&sc);
+    auto sc = std::make_shared<submitblock_StateCatcher>(block.GetHash());
+    RegisterSharedValidationInterface(sc);
     bool accepted = ProcessNewBlock(Params(), blockptr, /* fForceProcessing */ true, /* fNewBlock */ &new_block);
-    UnregisterValidationInterface(&sc);
+    UnregisterSharedValidationInterface(sc);
     if (!new_block && accepted) {
```

Comment:
An alternative is to call `SyncWithValidationInterfaceQueue` here.

Reply:
Alternative to what? The notification we are listening to is never put in the queue.

Reply:
I mean that after `UnregisterValidationInterface + SyncWithValidationInterfaceQueue` there is no pointer to `sc` in validation queue.

Reply:
How does `SyncWithValidationInterfaceQueue` prevent new notifications from being scheduled for `sc`?

Reply:
New notifications won't be scheduled because the `UnregisterValidationInterface` above. Only problem is a concurrent notification being called, so `SyncWithValidationInterfaceQueue` ensures that case doesn't happen.

Reply:
Now I understand. Yeah, correct.